### ISSUE TITLE
Improve documentation

### ISSUE THREAD

**Issue by gerw** [ownrecipes/OwnRecipes#111]
I recently set up OwnRecipes and I stumbled on one point: Everything was running, but the browser was not able to connect to the api. I followed the suggestions in `Troubleshooting`, but it didn't help.

The network panel in firefox wasn't showing any connection to the API. Finally, I launched chromium and it did show failed connections (with a CORS error). I had entered the URL of the API without `http://` and this was the problem, i.e., I changed
````
REACT_APP_API_URL=raspberrypi:8000
REACT_APP_ADMIN_URL=raspberrypi:8000/admin
````
to
````
REACT_APP_API_URL=http://raspberrypi:8000
REACT_APP_ADMIN_URL=http://raspberrypi:8000/admin
````
in the env file.

Note that `http://` is present in `Running_the_App.md` and `Running_the_App.md` (I somehow missed it), but it is missing in the comments of the sample `.env.docker.production.web.localhost`.

I suggest that
* the comments in `.env.docker.production.web.localhost` should also contain the protocol `http://`
* to add some comment in `Troubleshooting.md` which points to these variables.

**Comment by sepulzera** [comment]
Hey gerw! Glad to hear that you could get it up and running after all!

Your first suggestion about the `.env.docker.production.web.localhost` is correct. Thanks!

About your second suggestion: The `Troubleshooting.md` does point to the var `REACT_APP_API_URL`: https://github.com/ownrecipes/OwnRecipes/blob/master/docs/Troubleshooting.md#browser-can-not-connect-to-the-api. I can see that the wall of text can be quite overwhelming. Perhaps you could improve it (i .e. make a PR, or just write it here), to make it easier to get all the needed information? As deep-insighted maintainer I have sometimes a hard time doing so. :-)

Btw: How did you like the docs overall?

**Comment by gerw** [comment]
Mmh, yes, `Troubleshooting.md` mentions `REACT_APP_API_URL`, but I thought, that I got the URL right. I am not sure, but could my error (of forgetting `http://`) could be a common error? Then, one could point out, that the `REACT_APP_API_URL` should contain `http://` or `https://`. Or should the `web` part of OwnRecipes check that `http://` or `https://` is at the beginning of that URL and throw an error otherwise??

I think the docs are quite good. I ended up with the following setting:
* Compile the `web` part on the laptop
* Running `api` in a docker container on an raspberrypi 4B
* I have already a `mariadb` and an `apache` running, i.e., I reuse the database and use `apache` as a reverse proxy.

I think that for me it would have been a little bit easier, if `Running_the_App.md` would have a small section at the beginning, which describes that only `api` is needed on the server (and `web` can be run at a different computer). But to be fair, this is already contained in `Running_the_App_Tricks.md`...

For building the `web` part, I used the following dockerfile:
```Dockerfile
FROM node:16.15-slim

WORKDIR /code

# Install app dependencies
COPY package.json ./
COPY package-lock.json ./
RUN npm install --no-audit

COPY . ./

ENTRYPOINT npx react-inject-env set && npm run build && rm -rf ./public-ui/* && cp -r ./build/* ./public-ui/
```

the `Dockerfile-release` did not work for me, but again, this might be some mistake from my side.


Finally, one point is left that I did not figure out. Suppose that my instance is available at two URLs, e.g., `ownrecipe.local` (from LAN) and `ownrecipe.gerw.someddns.org`. In this situation, I am not able to get the connection to the API right, because in both cases, `REACT_APP_API_URL` needs to be different? Is it possible to handle this situation?

**Comment by sepulzera** [comment]
> I think that for me it would have been a little bit easier, if Running_the_App.md would have a small section at the beginning, which describes that only api is needed on the server (and web can be run at a different computer). But to be fair, this is already contained in Running_the_App_Tricks.md...

Well, it is a common 3-tier-architecture and thus splitting that up on different devices is possible. But I doubt it is a common use case for such a small app like OwnRecipes... it makes things more complicated, without any improvement (at least none I can think of).

> I ended up with the following setting:
>     * Compile the web part on the laptop
>     * Running api in a docker container on an raspberrypi 4B
> For building the web part, I used the following dockerfile:
> [...]
> the Dockerfile-release did not work for me, but again, this might be some mistake from my side.

So, you set up the api on your Raspberry Pi, but opted to run the web part on your laptop? If you already use docker for the api on your server, why not just use it for web, too?

Also, if I understood it right, you build your custom docker image for web and then host it locally on your laptop? That seems overly complicated? You could just use the official docker image, or run it in dev. 🤔 Now it seems you are mixing release and dev mode.
PS: You are calling `npx react-inject-env`, but not installing it (your file misses the command [`RUN npm install --save-dev react-inject-env@2.1.0`](https://github.com/ownrecipes/ownrecipes-web/blob/master/Dockerfile-release#L6)).

> Finally, one point is left that I did not figure out. Suppose that my instance is available at two URLs, e.g., ownrecipe.local (from LAN) and ownrecipe.gerw.someddns.org. In this situation, I am not able to get the connection to the API right, because in both cases, REACT_APP_API_URL needs to be different? Is it possible to handle this situation?

Well yeah, I can feel you. 😅
As it seems that your instance is publicly available, you should provide SSL (https), f. e. get a cert from Lets Encrypt. The cert would be valid for `ownrecipe.gerw.someddns.org`, and not `ownrecipe.local`.

At this point I was triggered, as I wanted to use https via intranet, too, and without cert warnings. My solution was to use `ownrecipe.gerw.someddns.org` from my intranet devices, too. Usually, then your traffic would route via internet, what is fairly useless (though with https still secure). Thus two possible solutions for that:
1. Add `ownrecipe.gerw.someddns.org` to the local devices hosts file. That isn't sexy, and difficult for mobile devices.
2. Set up your custom dns server `unbound` on your Raspberry Pi, and let it respond to `ownrecipe.gerw.someddns.org` with its static intranet ip (e.g. `192.168.178.xxx`). Hosting your own dns server brings some more improvements/possibilities: secured dns (DNSoverTLS or DNSoverHTTPS), reduced lookup-times for your regularily used addresses/sites, ads filter, and more. All in all, more sexy than the other solution, but requires quite some time to get it running.

If none of this is appealing, you could just run a second instance of web. Shouldn't have a big impact.

**Comment by sepulzera** [comment]
> But at least a warning would not break anything?

I will have a look.

>>  So, you set up the api on your Raspberry Pi, but opted to run the web part on your laptop? If you already use docker for the api on your server, why not just use it for web, too?
>
> I think that I had some troubles with docker (see below) and also the docs suggested to run web on a different machine.

I guess you are referring to [Running the App: Tricks - Single-board computer](https://github.com/ownrecipes/OwnRecipes/blob/master/docs/Running_the_App_Tricks.md#single-board-computer)?

It says to just build the web locally and run everything without docker. As you are already using docker, there would be no point to not use that for web as well... at least if api works (performance wise), I assume that web will do, too. But there is nothing wrong in using it the way you did, I guess... only just that maintaining (updating) it gets more complicated.

> No, I use docker on my laptop to create public-ui which I copied to the raspberrypi.

I never thought about using it that way. 😅
Now I am wondering if it is something that I should promote, e. g. include a Dockerfile-build for that. Just another file and more documentation... 🤔

> Concerning the Dockerfiles: If I use Dockerfile-release of the [web], I get an error, because there is no directory/file build as mentioned in Dockerfile-release.

Yes. Because you are mixing release and dev mode. That is why the file is called `Dockerfile-release` ;-). You would have to follow the docs to [Creating a release](https://github.com/ownrecipes/OwnRecipes/blob/master/docs/Creating_a_Release.md#ownrecipes-web) then.

**Comment by gerw** [comment]
> I guess you are referring to [Running the App: Tricks - Single-board computer](https://github.com/ownrecipes/OwnRecipes/blob/master/docs/Running_the_App_Tricks.md#single-board-computer)?

Yes, and this is what I extracted from this 😅

> I never thought about using it that way. 😅 Now I am wondering if it is something that I should promote, e. g. include a Dockerfile-build for that. Just another file and more documentation... 🤔

If I understand it correctly, this seems to be what my Dockerfile is doing. This seems to be a quite lightweight solution, since I can reuse my existing web server and database.

Now, I am also understanding what the hosted docker image of `web`: It already contains the compiled frontend, it injects the environment variables and copies the files to the desired location. (I was not able to use this docker image, since it was not provided for `arm` architecture. And running the docker of `web` on the raspberrypi was eating to much memory.)

> > Concerning the Dockerfiles: If I use Dockerfile-release of the [web], I get an error, because there is no directory/file build as mentioned in Dockerfile-release.
> 
> Yes. Because you are mixing release and dev mode. That is why the file is called `Dockerfile-release` ;-). You would have to follow the docs to [Creating a release](https://github.com/ownrecipes/OwnRecipes/blob/master/docs/Creating_a_Release.md#ownrecipes-web) then.

Ok, I see.

**Comment by gerw** [comment]
Is `RUN npm install --save-dev react-inject-env@2.1.0` really needed in the Dockerfile?

For extracting the build: One could use a docker volume with the `-v` option of `docker create` to mount the `build` dir from the container to an external `public-ui` repository (if it is already present, it should be cleaned).

**Comment by sepulzera** [comment]
> Is RUN npm install --save-dev react-inject-env@2.1.0 really needed in the Dockerfile?

Well, it is a docker image being built after all... so I thought it would be more complete then. But yeah, if not running it as container, then it is of no use. 🤷‍♂️

> For extracting the build: One could use a docker volume with the -v option of docker create to mount the build dir from the container to an external public-ui repository (if it is already present, it should be cleaned).

I actually tried that, but failed because of my limited docker experience. :-/
Could you please elaborate on that?

**Comment by sepulzera** [comment]
After more testing, I realized that both didn't work flawlessly. creat-react-app needs the env files, to properly replace the env-variables (see `ownrecipes-web/.env`). Only passing them via `docker run ... --env-file` would result in the env-variables not being replaced by create-react-app.

In order to get a fully functional build, I had to bypass the dockerignore, and create the docker image with all the files (+ env files), and build that. There is now a [script to build the web with docker](https://github.com/ownrecipes/ownrecipes-web/blob/development/docs/scripts/docker_build.sh).

And i opted for `docker create` over `docker run`. There is no need to actually run the docker container.

Thanks for your feedback and help!
